pycalphad/core/eqsolver.pyx

```diff
@@ -447,66 +368,51 @@ def _solve_eq_at_conditions(dbf, comps, properties, phase_records, conds_keys, v
         allow_negative_fractions = False
         for cur_iter in range(MAX_SOLVE_ITERATIONS):
             # print('CUR_ITER:', cur_iter)
-            phases = list(prop_Phase_values[it.multi_index])
-            if '' in phases:
-                old_phase_length = phases.index('')
-            else:
-                old_phase_length = -1
             if cur_iter > 0.8 * MAX_SOLVE_ITERATIONS:
                 allow_negative_fractions = False
-            remove_degenerate_phases(prop_Phase_values[it.multi_index], prop_X_values[it.multi_index],
-                                     prop_Y_values[it.multi_index], prop_NP_values[it.multi_index], allow_negative_fractions)
-            phases = list(prop_Phase_values[it.multi_index])
-            if '' in phases:
-                new_phase_length = phases.index('')
-            else:
-                new_phase_length = -1
-            # Are there removed phases?
-            if '' in phases:
-                num_phases = phases.index('')
-            else:
-                num_phases = len(phases)
+            if cur_iter > 0 and cur_iter % 5 == 0:
+                if cur_iter == 0:
+                    minimum_df = -10
+                else:
+                    minimum_df = 0
+                changed_phases |= add_new_phases(composition_sets, phase_records, current_grid, chemical_potentials, minimum_df, verbose)
+            changed_phases |= remove_degenerate_phases(composition_sets, allow_negative_fractions, verbose)
+            num_phases = len(composition_sets)
+            total_dof = sum([compset.phase_record.phase_dof for compset in composition_sets])
             if num_phases == 0:
-                raise ValueError('Zero phases are left in the system', cur_conds)
-            zero_dof = np.all(
-                (prop_Y_values[it.multi_index] == 1.) | np.isnan(prop_Y_values[it.multi_index]))
-            if (num_phases == 1) and zero_dof:
+                print('Zero phases are left in the system: {}'.format(cur_conds))
+                converged = False
```

Comment:
Good catch, but I'm going to argue for leaving it as-is because the control flow isn't obvious and this makes our intentions clearer.